### PR TITLE
[Core][Frontend] Add weight version tagging for RL rollouts

### PR DESCRIPTION
**Pull request by ShuoleiWang**
## Purpose

Part of #48306, this PR implements the engine-level externally managed weight-version metadata and query/update APIs from Section 2.2. It builds on the weight-update lifecycle introduced in #39212.

For online RL inference:

- Each `EngineCore` starts with `weight_version="default"`.
- Callers may pass an opaque string such as `"step-42"` to `finish_weight_update`. The supplied version is published only after the `finish_weight_update` collective RPC completes successfully. Calling `update_weights` does not change it, a failed finish does not publish it, and omitting it preserves the current version.
- The version can also be changed independently through `LLM.update_weight_version()`, `AsyncLLM.update_weight_version()`, or the Python frontend's dev-mode `POST /update_weight_version` endpoint. These APIs do not transfer weights or run the weight-update lifecycle.
- The current version can be read through `LLM.get_weight_version()`, `AsyncLLM.get_weight_version()`, or the Python frontend's dev-mode `GET /weight_info` endpoint.

The synchronous query path is:

```text
LLM.get_weight_version()
  -> LLMEngine.get_weight_version()
  -> EngineCoreClient.get_weight_version()
  -> EngineCore.get_weight_version()
```

The asynchronous query path is:

```text
AsyncLLM.get_weight_version()
  -> EngineCoreClient.get_weight_version_async()
  -> EngineCore.get_weight_version()
```

Version updates use the corresponding setter path:

```text
LLM.update_weight_version()
  -> LLMEngine.set_weight_version()
  -> EngineCoreClient.set_weight_version()
  -> EngineCore.set_weight_version()

AsyncLLM.update_weight_version()
  -> EngineCoreClient.set_weight_version_async()
  -> EngineCore.set_weight_version()
```

With `VLLM_SERVER_DEV_MODE=1`:

- `POST /finish_weight_update` accepts an optional `{"weight_version": "step-42"}` body and publishes it only after a successful finish.
- `POST /update_weight_version` accepts `{"new_version": "step-42"}` and changes only the version metadata.
- `GET /weight_info` calls `EngineClient.get_weight_version()` and returns `{"weight_version": "<current version>"}` through the same asynchronous path.

This PR intentionally does not bind a version to `Request`, `EngineCoreOutput`, or `RequestOutput`, and it does not add weight-version metadata to Python or Rust token-in/token-out responses. Therefore, `/weight_info` reports the current engine version at query time; it does not identify which weight version or versions produced the tokens of a particular rollout.

The version metadata belongs to one `EngineCore` and resets to `"default"` when that core restarts. This PR does not add persistence, the RFC's multi-replica `(dp_rank, version)` identity, distributed rollback, or router-side version selection. It does not change model execution, logits, token selection, scheduling policy, request handling, or the underlying weight-transfer mechanics.

## Test Plan

- Extend the existing real-`LLM` weight-transfer tests to verify that the version starts at `"default"`, remains unchanged before finish, becomes `"step-42"` after a successful finish, and can be changed independently to `"manual-version"`.
- Extend the existing trainer-client tests to verify that the HTTP client forwards `weight_version` in the finish request and that the Ray client waits for all actors to finish before publishing the version.
- Run focused Python regressions, E2E test collection, pre-commit, mypy, and diff checks.

## Test Result

Relevant Python TITO, protocol, and output regressions (`8 passed`):

```bash
.venv/bin/pytest -q tests/entrypoints/scale_out/token_in_token_out/test_generate_stream.py::test_serve_tokens_skips_mm_cache_for_remote_engine_execution tests/entrypoints/scale_out/token_in_token_out/test_generate_stream.py::test_stream_basic tests/entrypoints/scale_out/token_in_token_out/test_protocol.py tests/test_outputs.py
```

The real-engine and server suites collect successfully:

```bash
.venv/bin/python -m pytest --collect-only -q tests/entrypoints/weight_transfer/test_weight_transfer_llm.py tests/entrypoints/scale_out/token_in_token_out/test_serving_tokens.py
```

- Collection result: `14 tests collected`.
- The real-engine weight-transfer tests and `RemoteOpenAIServer` tests were not completed locally and will be validated by PR CI.
- `tests/distributed/test_weight_transfer.py::TestTrainerClients` was not executed locally because Ray is not installed.
- The OpenAPI schema suite was not executed locally because Schemathesis is not installed.
- The final 16-file PR diff passed `git diff --check` and contains no Rust changes.

### AI assistance disclosure

OpenAI Codex was used for code exploration, implementation assistance, and code review. I reviewed and understood every changed line and ran the tests reported above.

### CODE REVIEW

File: docs/training/async_rl.md [hunk 1]
```diff
@@ -59,6 +61,16 @@ A typical async RL loop with weight syncing looks like this:
 
 The key insight is that requests paused with `mode="keep"` will produce tokens from the **old** weights before the pause and tokens from the **new** weights after resume. The `clear_cache` parameter controls whether the KV cache is invalidated during the pause. When `clear_cache=True`, previously cached key-value entries are discarded, so all tokens generated after resume will be computed entirely with the new weights. When `clear_cache=False`, existing KV cache entries are retained, meaning some tokens in context may still reflect the old weights (stale KV cache).
 
+## Weight Version Tagging
```
Comment (aoshen02) [inline docs/training/async_rl.md]: Can be simpler, no need to have a separate paragraph.

Comment (ShuoleiWang) [reply docs/training/async_rl.md]: OK, Thanks!

File: tests/entrypoints/serve/dev/test_rlhf.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+
+from unittest.mock import AsyncMock, MagicMock
+
+from fastapi import FastAPI
+from fastapi.testclient import TestClient
+
+from vllm.engine.protocol import EngineClient
+from vllm.entrypoints.serve.dev.rlhf.api_router import attach_router
+
```
Comment (aoshen02) [inline tests/entrypoints/serve/dev/test_rlhf.py]: no need to have a test file here.

File: tests/v1/engine/test_weight_version.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+from types import SimpleNamespace
+from unittest.mock import MagicMock
+
+import pytest
+
+from vllm.v1.engine.core import EngineCore
+
+
+def _make_engine_core() -> EngineCore:
+    engine_core = object.__new__(EngineCore)
```
Comment (aoshen02) [inline tests/v1/engine/test_weight_version.py]: we only need to add some e2e test, ai-generated unit test is not recommended to add if without any human involved

Comment (ShuoleiWang) [reply tests/v1/engine/test_weight_version.py]: I see, I'll remove them.

Comment (ShuoleiWang) [reply tests/v1/engine/test_weight_version.py]: I have removed some unit tests and assertions. I kept minimal coverage in 2 existing test suites: the **real-LLM weight-transfer integration test** verifies that the version advances only after a successful finish, and the **RemoteOpenAIServer** tests cover /weight_info and streaming/non-streaming response metadata.

File: vllm/v1/request.py [hunk 4]
```diff
@@ -97,6 +97,9 @@ def __init__(
         self.status = RequestStatus.WAITING
         self.events: list[EngineCoreEvent] = []
         self.stop_reason: int | str | None = None
+        # Target-policy weight generation bound when this request is admitted
+        # to the scheduler. EngineCore sets this before Scheduler.add_request.
```
Comment (aoshen02) [inline vllm/v1/request.py]: Why we need to bind model version to the request? Since one request may go through several model version, if we want to late request have one, we might need to make it as a list I guess, but it would be very complicated?

Comment (ShuoleiWang) [reply vllm/v1/request.py]: Thanks for your comment. It is becuase of the text in #48306 Section 2.2:

> “In-flight requests bind to the **version at admission time** (locked when the request enters the scheduler queue).”

I understand that a request may span multiple model executions. For example, tokens generated before `pause_generation(mode="keep")` may use version x, while tokens generated after the weight update and resume may use version x+1. This scalar tag is not intended to represent the per-token execution history. Do you have any suggestions?

Comment (aoshen02) [reply vllm/v1/request.py]: Let's just remove it for now.

File: vllm/outputs.py [hunk 5]
```diff
@@ -193,7 +199,8 @@ def __repr__(self) -> str:
             f"metrics={self.metrics}, "
             f"lora_request={self.lora_request}, "
             f"num_cached_tokens={self.num_cached_tokens}, "
-            f"num_cache_creation_tokens={self.num_cache_creation_tokens})"
```
Comment (aoshen02) [inline vllm/outputs.py:196]: do we need to add this to request output?

File: vllm/v1/engine/output_processor.py [hunk 6]
```diff
@@ -631,6 +634,8 @@ def process_outputs(
             stop_reason = engine_core_output.stop_reason
             kv_transfer_params = engine_core_output.kv_transfer_params
             ec_transfer_params = engine_core_output.ec_transfer_params
+            if req_state.weight_version is None:
```
Comment (aoshen02) [inline vllm/v1/engine/output_processor.py]: Here as well

File: vllm/v1/engine/core_client.py [hunk 7]
```diff
@@ -169,9 +169,15 @@ def is_sleeping(self) -> bool:
     def execute_dummy_batch(self) -> None:
         raise NotImplementedError
 
+    def get_weight_version(self) -> int:
+        raise NotImplementedError
+
     async def execute_dummy_batch_async(self) -> None:
         raise NotImplementedError
 
+    async def get_weight_version_async(self) -> int:
```
Comment (aoshen02) [inline vllm/v1/engine/core_client.py]: why we need a async version here.

**Review by Isotr0py — APPROVED** [review]
